src/os_common/os_fhandle.c

```diff
@@ -355,6 +357,101 @@ __wt_close(WT_SESSION_IMPL *session, WT_FH **fhp)
 	return (__handle_close(session, fh, true));
 }
 
+/*
+ * __wt_fsync_background_chk --
+ *	Return if background fsync is supported.
+ */
+bool
+__wt_fsync_background_chk(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_FH *fh, *fhnext;
+	WT_FILE_HANDLE *handle;
+	bool supported;
+
+	conn = S2C(session);
+	supported = true;
+	__wt_spin_lock(session, &conn->fh_lock);
+	/*
+	 * Look for the first data file handle and see if
+	 * the fsync nowait function is supported.
+	 */
+	TAILQ_FOREACH_SAFE(fh, &conn->fhqh, q, fhnext) {
+		handle = fh->handle;
+		if (fh->file_type != WT_FS_OPEN_FILE_TYPE_DATA)
+			continue;
+		/*
+		 * If we don't have a function, return false, otherwise
+		 * return true. In any case, we are done with the loop.
+		 */
+		if (handle->fh_sync_nowait == NULL)
+			supported = false;
+		break;
+	}
+	__wt_spin_unlock(session, &conn->fh_lock);
+	return (supported);
+}
+
+/*
+ * __wt_fsync_background --
+ *	POSIX background fsync for all dirty file handles.
+ */
+int
+__wt_fsync_background(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_FH *fh;
+	WT_FILE_HANDLE *handle;
+	uint64_t now;
+
+	conn = S2C(session);
+	__wt_spin_lock(session, &conn->fh_lock);
+	TAILQ_FOREACH(fh, &conn->fhqh, q) {
+		handle = fh->handle;
+		WT_STAT_CONN_INCR(session, fsync_all_fh_total);
+		if (handle->fh_sync_nowait == NULL ||
+		    fh->written < WT_CAPACITY_FILE_THRESHOLD)
+			continue;
+		/* Only sync data files. */
+		if (fh->file_type != WT_FS_OPEN_FILE_TYPE_DATA)
+			continue;
+		now = __wt_clock(session);
+		if (fh->last_sync == 0 ||
+		    WT_CLOCKDIFF_SEC(now, fh->last_sync) > 0) {
+			/*
+			 * Increment our ref count on the current and next
+			 * handle. That way both are guaranteed valid when we
+			 * lock again after the fsync.
+			 */
+			++fh->ref;
```

Comment:
I think there's still an open question on this code:

> What if there's one other reference to the handle, we increment the reference count, then the other reference is closed (and our reference prevents the handle from being really closed), and then we decrement our reference count. Won't that leave a handle with a reference count of 0?

Reply:
Thanks, you're right.  I just pushed a fix.